### PR TITLE
feat: cache start system and solver in HomotopyContinuation interface

### PR DESCRIPTION
**Pull request by AayushSabharwal**
## Checklist

- [ ] Appropriate tests were added
- [ ] Any code changes were done in a way that does not break public API
- [ ] All documentation related to code changes were updated
- [ ] The new code follows the
  [contributor guidelines](https://github.com/SciML/.github/blob/master/CONTRIBUTING.md), in particular the [SciML Style Guide](https://github.com/SciML/SciMLStyle) and
  [COLPRAC](https://github.com/SciML/COLPRAC).
- [ ] Any new documentation only uses public API

### CODE REVIEW

**Comment by AayushSabharwal** [comment]
I realized polynomial checking doesn't take into account observed variables, so we don't report a nice error.

File: test/extensions/homotopy_continuation.jl [hunk 1]
```diff
@@ -61,20 +61,11 @@ end
...
-    prob = HomotopyContinuationProblem(sys, [])
-    sol = solve(prob; threading = false)
-    @test SciMLBase.successful_retcode(sol)
-end
-
 @testset "Polynomial check and warnings" begin
     @variables x = 1.0
-    @parameters n = 4
+    @parameters n::Integer = 4
     @mtkbuild sys = NonlinearSystem([x^n + x^2 - 1 ~ 0])
```
Comment (isaacsas) [inline test/extensions/homotopy_continuation.jl:67]: No supporting integer exponents is a pretty big limitation. It means that very common expressions in bio/pharma modeling, Hill expressions, would not be usable here (i.e. terms like ` x^n / (K + x^n)` where `n` is an integer parameter). Is there no way to handle them? They should, ultimately, reduce down to polynomials once the denominators are removed and the integer exponent values are substituted in.

Comment (AayushSabharwal) [reply test/extensions/homotopy_continuation.jl:67]: Interesting. We can't cache the start system with parametric exponents. @ChrisRackauckas I guess we just have to do the caching optionally?

Comment (ChrisRackauckas) [reply test/extensions/homotopy_continuation.jl:67]: Why can't the hill functions in this case use `@constant`?

Comment (AayushSabharwal) [reply test/extensions/homotopy_continuation.jl:67]: It's also worth noting that pure HomotopyContinuation.jl also doesn't allow parametric exponents

Comment (AayushSabharwal) [reply test/extensions/homotopy_continuation.jl]: Parametric exponents are re-allowed with a warning

**Comment by AayushSabharwal** [comment]
Extension test failures are BifurcationKit failing to precompile

**Comment by ChrisRackauckas** [comment]
But this would be in extension tests. Just flip the order so we can see it pass.

File: test/extensions/homotopy_continuation.jl (hunk 1 above)
Comment (isaacsas) [reply test/extensions/homotopy_continuation.jl]: Great, thanks.

**Pull request merged**